pkg/operator/controller/dns_status.go

```diff
@@ -0,0 +1,146 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// syncDNSStatus computes the current status of dns and
+// updates status upon any changes since last sync.
+func (r *reconciler) syncDNSStatus(dns *operatorv1.DNS, clusterIP, clusterDomain string, ds *appsv1.DaemonSet) error {
+	updated := dns.DeepCopy()
+	updated.Status.ClusterIP = clusterIP
+	updated.Status.ClusterDomain = clusterDomain
+	updated.Status.Conditions = computeDNSStatusConditions(updated.Status.Conditions, clusterIP, ds)
+	if !dnsStatusesEqual(updated.Status, dns.Status) {
+		if err := r.client.Status().Update(context.TODO(), updated); err != nil {
+			return fmt.Errorf("failed to update dns status: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// computeDNSStatusConditions computes dns status conditions
+// based on the status of ds.
+func computeDNSStatusConditions(conditions []operatorv1.OperatorCondition, clusterIP string, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
+	conditions = computeDNSDegradedCondition(conditions, clusterIP)
+	conditions = computeDNSProgressingCondition(conditions, ds)
+	conditions = computeDNSAvailableCondition(conditions, ds)
+
```

Comment:
`computeDNSStatusConditions` does not depend on old status conditions so we don't have to pass this as an argument?
May be we could do:
```
func computeDNSStatusConditions(ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
        conditions := []operatorv1.OperatorCondition{}
	conditions = append(conditions, computeDNSDegradedCondition(clusterIP))
	conditions = append(conditions, computeDNSProgressingCondition(ds))
	conditions = append(conditions, computeDNSAvailableCondition(ds))
        return conditions
}
```
Each of the `computeDNSConditions` methods could return a condition and we no longer need the `setDNSStatusCondition` method.

Reply:
`computeDNSFooCondition` needs the old condition as input so that it can return the old condition with the old last transition time if the `Foo` condition did not change.

Reply:
On further thought, we could do something very close to Ravi's suggestion:
```go
func computeDNSStatusConditions(conditions []configv1.ClusterOperatorStatusCondition, ds *appsv1.DaemonSet) []operatorv1.OperatorCondition {
	conditions = setOperatorStatusCondition(conditions, computeDNSDegradedCondition(clusterIP))
	conditions = setOperatorStatusCondition(conditions, computeDNSProgressingCondition(ds))
	conditions = setOperatorStatusCondition(conditions, computeDNSAvailableCondition(ds))
        return conditions
}
```
I kind of like that, and it should correctly preserve or as appropriate update `LastTransitionTime`.